### PR TITLE
perf: refactor URLInfo to not use ParseResult inside

### PR DESCRIPTION
**Pull request by Suor**
This permites URLInfo construction from parts without building/reparsing
string, which makes it around 20% faster. Code is also simpler and a bit
shorter now.

Other improvements:
- added URLInfo.replace(path=...)
- URLInfo._path now stringifies properly
- protected against passing path not starting with /
- do not allow query, params, fragments and passwords in urls (they were
silently ignored previously)

* [x] Have you followed the guidelines in our
      [Contributing document](https://dvc.org/doc/user-guide/contributing/core)?

* [x] Does your PR affect documented changes or does it add new functionality
      that should be documented? If yes, have you created a PR for
      [dvc.org](https://github.com/iterative/dvc.org) documenting it or at
      least opened an issue for it? If so, please add a link to it.

-----

### CODE REVIEW

**Comment by Suor** [comment]
Here is a bench script I used:
```python
from funcy import chain
from dvc.path_info import URLInfo, CloudURLInfo


urls = [URLInfo('ssh://host.com/a/b/%s' % x) for x in range(2000)]
curls = [CloudURLInfo('ssh://host.com/a/b/%s' % x) for x in range(2000)]

urls_set = set(urls) | set(curls)
for url in chain(urls, curls):
    urls_set.add(url)

x = 0
for u in urls:
    for u2 in urls[:50]:
        x += u == u2

add = [u / "x" for u in urls]
cadd = [u / "x" for u in curls]
```

File: dvc/path_info.py [hunk 1]
```diff
@@ -112,52 +111,80 @@ class PosixPathInfo(PathInfo, pathlib.PurePosixPath):
...
 
 
 class URLInfo(object):
     DEFAULT_PORTS = {"http": 80, "https": 443, "ssh": 22, "hdfs": 0}
 
     def __init__(self, url):
-        self.parsed = urlparse(url)
-        assert self.parsed.scheme != "remote"
+        p = urlparse(url)
+        assert not p.query and not p.params and not p.fragment
```
Comment (mroutis) [inline dvc/path_info.py]: why do we need these assertions?

Comment (Suor) [reply dvc/path_info.py]: We don't store that, so in the absence of asserts:
```python
URLInfo("http://host.com/a/b?q=1&p=2#ref") == URLInfo("http://host.com/a/b")
``` 
Which might surprise people at some point, i.e. lead to an "entertaining" debugging session.

**Comment by Suor** [comment]
I wonder what you guys would say when actual metaprogramming would be involved :)

**Comment by Suor** [comment]
@efiop may you take a look?

**Pull request merged**